base10.go

```diff
@@ -0,0 +1,132 @@
+package uint256
+
+import (
+	"strconv"
+	"strings"
+)
+
+const twoPow256Sub1 = "115792089237316195423570985008687907853269984665640564039457584007913129639935"
+const twoPow128 = "340282366920938463463374607431768211456"
+const twoPow64 = "18446744073709551616"
+
+func (z *Int) Base10() string {
+	return z.ToBig().String()
+}
+
+// SetString implements a subset of (*big.Int).SetString
+// ok will be true iff i == nil
+func (z *Int) SetString(s string, base int) (i *Int, ok bool) {
+	switch base {
+	case 0:
+		if strings.HasPrefix(s, "0x") {
+			err := z.fromHex(s)
+			if err != nil {
+				return nil, false
+			}
+			return z, true
+		}
+		err := z.SetFromBase10(s)
+		if err != nil {
+			return nil, false
+		}
+		return z, true
+	case 10:
+		err := z.SetFromBase10(s)
+		if err != nil {
+			return nil, false
+		}
+		return z, true
+	case 16:
+		err := z.fromHex(s)
+		if err != nil {
+			return nil, false
+		}
+		return z, true
+	}
+	return nil, false
+}
+
+// FromBase10 is a convenience-constructor to create an Int from a
+// decimal (base 10) string. Numbers larger than 256 bits are not accepted.
+func FromBase10(hex string) (*Int, error) {
+	var z Int
+	if err := z.SetFromBase10(hex); err != nil {
+		return nil, err
+	}
+	return &z, nil
+}
+
+// SetFromBase10 sets z from the given string, interpreted as a decimal number.
+func (z *Int) SetFromBase10(s string) (err error) {
+	if len(s) < len(twoPow256Sub1) {
+		return z.fromBase10Long(s)
+	}
+	if len(s) == len(twoPow256Sub1) {
+		if s > twoPow256Sub1 {
+			return ErrBig256Range
+		}
+		return z.fromBase10Long(s)
+	}
+	return ErrBig256Range
+}
+
+var scaleTable10 [78]Int
+
+func init() {
+	for k := range scaleTable10 {
+		scaleTable10[k].Exp(NewInt(10), NewInt(uint64(k)))
+	}
+}
+
+// helper function to only ever be called via SetFromBase10
+// this function takes a string and chunks it up, calling ParseUint on it up to 5 times
+// these chunks are then multiplied by the proper power of 10, then added together.
+func (z *Int) fromBase10Long(bs string) error {
+	// first clear the input
+	z.Clear()
+	// if the input value is empty string, just do nothing. effectively, empty string sets to 0
+	if bs == "" {
+		return nil
+	}
+	// the maximum value of uint64 is 18446744073709551615, which is 20 characters
+	// one less means that a string of 19 9's is always within the uint64 limit
+	cutLength := 19
+	// cutStart tracks the current position of the string that we are in
+	cutStart := 0
+	// start iterating from 4 to 1. This is because the maximum value of uint256 is 78 characters,
+	// which can be divided into 5 integers of up to 19 characters.
+	// however, the last number will always be below 19 characters, so i=0 is dealt with as special case
+	for i := 4; i >= 1; i-- {
+		// check if the length of the string is larger than cutLength * i
+		if len(bs) >= (cutLength * i) {
```

Comment:
Reverse the check, so you can unindent the clause
```
if len(bs) < (cutLength * i){
    continue
}
```
Although, I wonder if it wouldn't make sense to start a better index. Something like this (might be off-by-one):
```
for i := len(bs) / 19; ...
```
Also, cutLength is a constant. It's ok to have it as a `var`, but I don't know, we could just also inline `19`. I don't know which I prefer.

Reply:
good catch on the loop, my bad! switched it to use continue and start at len(bs)/cutLength

I don't think it's off by one since if it's below 19, we would want it to be 0, and ignore the loop altogether, while otherwise, we are looking for the floor. if it's 19, it will perfectly match, and the stragglers case will get short circuited

cutLength is only used in this function - to me it makes more sense to leave it inside the function, unless other things start reusing that number 19.